source/audio/midiplayer.cpp

```diff
@@ -185,11 +185,29 @@ void MidiPlayer::run()
         if (!(event->isNoteOnOff() &&
               event->getChannel() == METRONOME_CHANNEL &&
               !myMetronomeEnabled) &&
-            !event->isTempoChange() && !event->isTrackEnd())
+            !event->isTempoChange() && 
+            !event->isTrackEnd() &&
+            !event->isVolumeChange())
         {
             device.sendMessage(event->getData());
         }
 
+        // get the channel and the correspondin player
+        int channel  = event->getChannel();
+        int player = getPlayerFromChannel(channel);
+        // if the channel corresponds to a valid player, set its maximum volume
+        if(player != -1)
+        {
+            device.setChannelMaxVolume(channel, myScore.getPlayers()[player].getMaxVolume());
+        }
+        // handle volume change events
+        // using device.setVolume() ensures that the maximum volume threshold
+        // is taken into consideration
+        if(event -> isVolumeChange())
```

Comment:
The formatting is a bit inconsistent here - should be `if (event->isVolumeChange())`